src/services/organizeImports.ts

```diff
@@ -0,0 +1,291 @@
+
+            if (kind1 !== kind2) {
+                return kind1 < kind2
+                    ? Comparison.LessThan
+                    : Comparison.GreaterThan;
+            }
+
+            // Note that we're using simple equality, retaining case-sensitivity.
+            if (name1 !== name2) {
+                return name1 < name2
+                    ? Comparison.LessThan
+                    : Comparison.GreaterThan;
+            }
+
+            return Comparison.EqualTo;
+        });
+
+        return sortedRecords.map(r => r.importDeclaration);
+
+        function createImportRecord(importDeclaration: ImportDeclaration) {
+            const specifier = importDeclaration.moduleSpecifier;
+            const name = getExternalModuleName(specifier);
+            if (name) {
+                const isRelative = isExternalModuleNameRelative(name);
+                return { importDeclaration, name, kind: isRelative ? ModuleNameKind.Relative : ModuleNameKind.NonRelative };
+            }
+
+            return { importDeclaration, name: specifier.getText(), kind: ModuleNameKind.Invalid };
+        }
+    }
+
+    function getExternalModuleName(specifier: Expression) {
+        return isStringLiteral(specifier) || isNoSubstitutionTemplateLiteral(specifier)
+            ? specifier.text
+            : undefined;
+    }
+
+    /**
+     * @param sortedImports a non-empty list of ImportDeclarations, sorted by module name.
+     */
+    function groupSortedImports(sortedImports: ReadonlyArray<ImportDeclaration>): ReadonlyArray<ReadonlyArray<ImportDeclaration>> {
+        Debug.assert(length(sortedImports) > 0);
+
+        const groups: ImportDeclaration[][] = [];
+
+        let groupName: string | undefined = getExternalModuleName(sortedImports[0].moduleSpecifier);
+        let group: ImportDeclaration[] = [];
+
+        for (const importDeclaration of sortedImports) {
+            const moduleName = getExternalModuleName(importDeclaration.moduleSpecifier);
+            if (moduleName && moduleName === groupName) {
+                group.push(importDeclaration);
+            }
+            else if (group.length) {
+                groups.push(group);
+
+                groupName = moduleName;
+                group = [importDeclaration];
+            }
+        }
+
+        if (group.length) {
+            groups.push(group);
+        }
+
+        return groups;
+    }
+
+    /* @internal */ // Internal for testing
+    /**
+     * @param sortedImports a list of ImportDeclarations, sorted by module name.
+     */
+    export function coalesceImports(sortedImports: ReadonlyArray<ImportDeclaration>) {
+        if (sortedImports.length === 0) {
+            return sortedImports;
+        }
+
+        const coalescedImports: ImportDeclaration[] = [];
+
+        const groupedImports = groupSortedImports(sortedImports);
+        for (const importGroup of groupedImports) {
+
+            let seenImportWithoutClause = false;

```

Comment:
Think this could be its own function:
```ts
function getImportParts(importGroup: ReadonlyArray<ImportDeclaration>
): {
    importWithoutClause: ImportDeclaration | undefined,
    defaultImports: ReadonlyArray<Identifier>,
    namespaceImports: ReadonlyArray<NamespaceImport>,
    namedImports: ReadonlyArray<NamedImports>,
} {
    let importWithoutClause: ImportDeclaration;
    const defaultImports: Identifier[] = [];
    const namespaceImports: NamespaceImport[] = [];
    const namedImports: NamedImports[] = [];

    for (const importDeclaration of importGroup) {
        if (importDeclaration.importClause === undefined) {
            // Only the first such import is interesting - the others are redundant.
            // Note: Unfortunately, we will lose trivia that was on this node.
            importWithoutClause = importDeclaration;
            continue;
        }

        const { name, namedBindings } = importDeclaration.importClause;

        if (name) {
            defaultImports.push(name);
        }

        if (namedBindings) {
            if (isNamespaceImport(namedBindings)) {
                namespaceImports.push(namedBindings);
            }
            else {
                namedImports.push(namedBindings);
            }
        }
    }

    return { importWithoutClause, defaultImports, namespaceImports, namedImports };
}
```
then:
```ts
const { importWithoutClause, defaultImports, namedImports, namespaceImports } = getImportParts(importGroup);
if (importWithoutClause) {
    coalescedImports.push(importWithoutClause);
}
...proceed as before...
```

Reply:
Sure.